src/backend/cuda/Array.cpp

```diff
@@ -220,46 +222,66 @@ namespace cuda
                     lock_bytes > getMaxBytes() ||
                     lock_buffers > getMaxBuffers();
 
-
                 // We eval in the following cases.
-                // 1. Too many bytes are locked up by JIT causing memory pressure.
-                // Too many bytes is assumed to be half of all bytes allocated so far.
-                // 2. Too many buffers in a nonlinear kernel cause param space overflow.
-                // Too many buffers comes out to be about 50 (51 including output).
-                // Too many buffers can occur in a tree of size 25 in the worst case scenario.
+                //
+                // 1. Too many bytes are locked up by JIT causing memory
+                //    pressure. Too many bytes is assumed to be half of all bytes
+                //    allocated so far.
+                //
+                // 2. Too many buffers in a nonlinear kernel cause param space
+                //    overflow. This happens when the number of nodes reaches 50
+                //    (51 including output). Too many buffers can occur in a tree
+                //    of size 25 in the worst case.
+                //
                 // TODO: Find better solution than the following emperical solution.
                 if (node->getHeight() > 25 || isBufferLimit) {
-
+                    // This is the size of the params that are passed by default
+                    constexpr int param_base_size = sizeof(Param<T>) + (4 * sizeof(uint));
+
+                    // This is the maximum size of the params that can be allowed by CUDA
+                    // NOTE: This number should have been (4096 - some_buffer_size) BUT
+                    // kernels who's kernel sizes come close to this value are not passing
+                    // and cuModuleLoadDataEx is failing with CUDA_ERROR_INVALID_IMAGE(200).
+                    // 35*sizeof(int) seems to be the magic number that passes all tests.
+                    // I have no idea why this is the case.
+                    constexpr int max_param_size = (4096 - (sizeof(Param<T>) + 35*sizeof(uint)));
```

Comment:
Why do you need to add another `sizeof(Param<T>)` ?

Reply:
We also pass in some additional metadata like the number of repeats along x and y. Number of dimensions etc. Would accounting for that make this magic constant less magical ?

Reply:
I was taking into account the output Array. The 35*sizeof(uint) was supposed to take that into account(should have been 4 instead of 35) but when we create a kernel that is within the range of the allowed size the invalid image error is thrown by ModuleLoadDataEx.